MRG: add denoising source separation (DSS)

drammock opened the pull request:
no tests or examples yet.  Comments / suggestions welcome.

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+        bias_thresh=None):
+    """Preprocess physiological data with denoising source separation (DSS)
+
+    Implementation follows the procedure described in S\"arel\"a & Valpola [1]
```
[1]_

Eric89GXL replied:
also, you should probably just use the `unicode` directly in the names

drammock replied:
Will gladly switch to unicode, I just didn't think you would allow it...  in expyfun there was some reason we had to maintain ASCII in the source files (something to do with Travis perhaps?)

Eric89GXL replied:
I think `sphinx` used to die but now it should be okay for mne-python at least, we use it elsewhere.

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+
+    References
+    ----------
+    [1] S\"arel\"a, Jaakko, and Valpola, Harri (2005). Denoising source
```
.. [1]

Eric89GXL replied:
(https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+    ----------
+    data : instance of Epochs
+        Data to be denoised.
+    data_n_keep, bias_n_keep : int
```
we do each param on a separate line

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+       will be discarded during decomposition of the data and bias function.
+       For example, if the largest eigenvalue is 3 and ``thresh=0.1``,
+       components with eigenvalues below 0.3 will be discarded.
+
```
Returns?

drammock replied:
Tricky question.  DSS will convert channels to (a possibly reduced number of) virtual channels, so returning a modified Epochs instance (whether in-place or copied) will require changing channels / channel names.  In light of that, does it make more sense to return just the DSS data?

Eric89GXL replied:
For now, yeah let's do that. We don't have a channel type in the FIFF format for DSS-type channels, and I don't think we want to make one for each such processing scheme. So just returning an array of the right shape is probably the cleanest option for now. There should also be an option to `return_operator` to get the spatial operator that was used in case people want to do some source localization.

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+    [2] de Cheveign\'e, Alain, and Simon, Jonathan Z. (2008). Denoising based
+    on spatial filtering. Journal of Neuroscience Methods, 171(2): 331-339.
+    """
+    # TODO: allow Raw as well as Epochs?
```
Any idea which will be used more often?

It doesn't seem that much more complicated to add `Raw` if I'm right that in the end it's just a spatial operation.

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+    n_trials, n_chans, n_times = data.shape
+    data = data.transpose((1, 2, 0)).reshape(n_chans, n_times * n_trials)
+    epochs._data = np.dot(dss_mat.T, data).reshape(dss_mat.shape[1], n_times,
+                                                   n_trials).transpose(2, 0, 1)
```
save yourself some transpose/reshape pain and use `np.einsum`, it's a cool function

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+    # TODO: apply picks before DSS?
+    evoked = epochs.average()
+    bias_cov = np.dot(evoked, evoked.T)
+    data_cov = np.sum([np.dot(trial, trial.T) for trial in epochs], axis=0)
```
better to use `compute_covariance` here

Eric89GXL replied:
... and note that it becomes pretty easy to support `Raw` because you just triage to `compute_covariance` for `Epochs` or `compute_raw_covariance` for `_BaseRaw`, and then modify your `einsum` at the end of the function to account for the number of dimensions.

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+    eigvec : array
+        2-dimensional array of eigenvectors.
+    """
+    eigval, eigvec = np.linalg.eig(cov)
```
cov is positive semidefinite so you should be able to use `eigh`

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+        eigval = eigval[:n_keep]
+        eigvec = eigvec[:, :n_keep]
+    if thresh is not None:
+        suprathresh = np.where(eigval / eigval.max() >= thresh)[0]
```
this is not always guaranteed to have any elements, should probably throw an error if `len(suprathresh) == 0` right?

Eric89GXL replied:
ahh this is a generic `_pca` function, so maybe throw the error in the calling function instead if necessary

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+    dss_mat = data_eigvec.dot(W).dot(bias_eigvec)
+    # matrix to normalize DSS dimensions
+    N = np.diag(np.sqrt(1 / np.diag(dss_mat.T.dot(data_cov).dot(dss_mat))))
+    return dss_mat.dot(N)
```
you shuold never need to dot with a `np.diag(x)` (where `x` is 1D), it's just a slower version of `* x` (by broadcasting)

Eric89GXL replied:
(or `* x[np.newaxis]` depending on which side you dot with)

Eric89GXL replied:
(whoops I mean `x[:, np.newaxis`)

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+    bias function to be the evoked response).
+    """
+    data_eigval, data_eigvec = _pca(data_cov, data_n_keep, data_thresh)
+    W = np.diag(np.sqrt(1 / data_eigval))  # whitening matrix
```
same thing, don't use dot with `np.diag(x)` there's a faster way

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+
+
+def dss(epochs, data_n_keep=None, bias_n_keep=None, data_thresh=None,
+        bias_thresh=None):
```
You probably don't want to allow `None` thresholding because you can end up with zero-value eigenvalues, right? Instead you could do e.g. `data_thresh='numerical'` which would use numerical precision, similar to `scipy.linalg.orth` [here] (https://github.com/scipy/scipy/blob/v0.14.0/scipy/linalg/decomp_svd.py#L181). If a user really wants to potentially break things they can go ahead and set it to zero themselves.

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,108 @@
+"""
+Created on Tue Apr 19 12:47:10 2016
+@author: drmccloy
+"""
```
reformat like other `mne-python` files

Eric89GXL commented:
Despite a bunch of minor gripes, LGTM :)

wmvanvliet commented:
Even in the sandbox, I would like new additions to have tests and at least one example script.

drammock commented:
@wmvanvliet I intend to add them (which I meant to imply with my initial PR comment):
> no tests or examples *yet*

drammock commented:
@Eric89GXL I *think* I hit all your comments.  Also added a couple example scripts, will add tests soon.

Eric89GXL commented on examples/preprocessing/plot_dss_array.py:
```diff
@@ -0,0 +1,54 @@
+#
+# License: BSD (3-clause)
+
+from __future__ import division
```
we try to avoid this, instead for consistency just do cast `float(x)` when doing division

Eric89GXL commented on examples/preprocessing/plot_dss_array.py:
```diff
@@ -0,0 +1,54 @@
+n_channels = 32
+pad = np.zeros(n_times // 3)
+signal_nsamps = n_times - 2 * pad.size
+signal = np.sin(2 * np.pi * np.arange(signal_nsamps) / signal_nsamps)
```
e.g. here

Eric89GXL commented on examples/preprocessing/plot_dss_epochs.py:
```diff
@@ -0,0 +1,58 @@
+    ax.yaxis.set_ticks_position('left')
+    ax.spines['top'].set_visible(False)
+    ax.spines['right'].set_visible(False)
+    _ = ax.plot(dat, linewidth=lw)
```
no need for `_ =`

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,140 @@
+                   bias_max_components, bias_thresh)
+    if return_data:
+        n_trials, n_chans, n_times = data.shape
+        dss_data = np.einsum('ij,hik->hjk', dss_mat, data)
```
people usually get confused with `einsum` so it might be worth adding a comment it's equivalent to `[np.dot(dss_mat, ep) for ep in data]`, even though it can be deduced from the indices

drammock replied:
will add note (though it's actually `np.array([np.dot(dss_mat.T, ep) for ep in data])`; you missed the `.T`)

Eric89GXL replied:
Ahh yes. Actually you should probably return `dss_mat.T` then because people will expect `dss_mat` to have shape `(n_dss, n_channels)`, i.e. I could immediately apply it with `np.dot(...)` to a single epoch. Also update the docstring that way.

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,140 @@
+    data_eigval, data_eigvec = _pca(data_cov, data_max_components, data_thresh)
+    W = np.sqrt(1 / data_eigval)  # diagonal of whitening matrix
+    # bias covariance projected into whitened PCA space of data channels
+    bias_cov_white = (W * data_eigvec).T.dot(bias_cov).dot(data_eigvec) * W
```
Did you do an e.g. `np.testing.assert_allclose` of this new version compared to your old one? I'm a little bit surprised that no `W[:, np.newaxis]` appears here but maybe that's because it implicily lives in the starting `(W * dat_eigvec).T`

drammock replied:
you're correct; I tested that `(W * A).T` is equivalent to `W[:, np.newaxis] * A.T`

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,140 @@
+    dss_mat = (W[np.newaxis, :] * data_eigvec).dot(bias_eigvec)
+    # normalize DSS dimensions
+    N = np.sqrt(1 / np.diag(dss_mat.T.dot(data_cov).dot(dss_mat)))
+    return N[np.newaxis, :] * dss_mat
```
you shouldn't need the `np.newaxis` here due to `numpy` broadcasting rules

drammock replied:
oops, that's cruft.

Eric89GXL commented:
Yep, looks like you got them all. Now we just need some tests and we should be good.

drammock commented:
@wmvanvliet DSS tests added, all of @Eric89GXL's comments addressed.

Eric89GXL commented on mne_sandbox/preprocessing/tests/test_dss.py:
```diff
@@ -0,0 +1,47 @@
+        assert_allclose(dss_trial1_comp1, sine, rtol=0, atol=0.2)
+    except AssertionError:
+        # maybe just 180 degree phase difference
+        assert_allclose(-1 * dss_trial1_comp1, sine, rtol=0, atol=0.2)
```
instead of doing this, you could explicitly sign flip with something like:
```
dss_trial_comp1 *= np.sign(np.dot(dss_trial1_comp1, sine))
assert_allclose(dss_trial1_comp1, sine, rtol=0, atol=0.2)
```

Eric89GXL commented:
LGTM but Travis has some weird errors

Eric89GXL commented on mne_sandbox/preprocessing/dss.py:
```diff
@@ -0,0 +1,145 @@
+# -*- coding: utf-8 -*-
+
+"""Denoising source separation"""
```
ahh, you should rename this module `_dss.py` because the import is getting screwed up by having a module `dss` and a `from .dss import dss` in `dss/__init__.py`

Eric89GXL commented:
Can you check on the coverage decrease?

Eric89GXL commented:
https://coveralls.io/builds/5906859/source?filename=mne_sandbox%2Fpreprocessing%2F_dss.py

drammock commented:
Sure, I'll have a look tomorrow.
On Apr 23, 2016 5:56 PM, "Eric Larson" <notifications@github.com> wrote:

> Can you check on the coverage decrease?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-sandbox/pull/9#issuecomment-213860504>
>

drammock commented:
@Eric89GXL I've addressed the coverage issue, and fixed one of the two Travis errors, but I'm not sure hwo to fix the other one (it appears to be trying to make a link with `ln` to a location inside an `.egg` file).  But the DSS tests pass on 3 of the 4 travis builds.

Eric89GXL commented:
Thanks @drammock
